scipy/sparse/_coo.py

```diff
@@ -109,112 +128,130 @@ def reshape(self, *args, **kwargs):
             else:
                 return self
 
-        nrows, ncols = self.shape
-
-        if order == 'C':
-            # Upcast to avoid overflows: the coo_array constructor
-            # below will downcast the results to a smaller dtype, if
-            # possible.
-            maxval = (ncols * max(0, nrows - 1) + max(0, ncols - 1))
-            dtype = self._get_index_dtype(maxval=maxval)
-
-            flat_indices = np.multiply(ncols, self.row, dtype=dtype) + self.col
-            new_row, new_col = divmod(flat_indices, shape[1])
-        elif order == 'F':
-            maxval = (nrows * max(0, ncols - 1) + max(0, nrows - 1))
-            dtype = self._get_index_dtype(maxval=maxval)
-
-            flat_indices = np.multiply(nrows, self.col, dtype=dtype) + self.row
-            new_col, new_row = divmod(flat_indices, shape[0])
+        # When reducing the number of dimensions, we need to be careful about
+        # index overflow. This is why we can't simply call
+        # `np.ravel_multi_index()` followed by `np.unravel_index()` here.
+        flat_indices = _ravel_indices(self.indices, self.shape, order=order)
+        if len(shape) == 2:
+            if order == 'C':
+                new_indices = divmod(flat_indices, shape[1])
+            else:
+                new_indices = divmod(flat_indices, shape[0])[::-1]
         else:
-            raise ValueError("'order' must be 'C' or 'F'")
+            new_indices = np.unravel_index(flat_indices, shape, order=order)
 
         # Handle copy here rather than passing on to the constructor so that no
-        # copy will be made of new_row and new_col regardless
+        # copy will be made of `new_indices` regardless.
```

Comment:
Admittedly, this comment doesn't make a lot of sense to me, but perhaps will if I dig deeper into the machinery.